src/global_distrib/mod_global_distrib_hosts_refresher.erl

```diff
@@ -0,0 +1,154 @@
+%%%==============================================================================
+%% Copyright 2018 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+
+
+-module(mod_global_distrib_hosts_refresher).
+-author("dominik.stanaszek@erlang-solutions.com").
+
+-behavior(gen_server).
+-behavior(gen_mod).
+-include("mongoose.hrl").
+
+%% API
+-export([start_link/1]).
+
+%% gen_server API
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2]).
+
+%% gen_mod API
+-export([deps/2, start/2, stop/1]).
+
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec start_link(Milliseconds :: non_neg_integer()) -> pid() | term().
+start_link(RefreshInterval) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [RefreshInterval], []).
+
+
+%%--------------------------------------------------------------------
+%% gen_mod callbacks
+%%--------------------------------------------------------------------
+
+-spec start(Host :: ejabberd:lserver(), Opts :: proplists:proplist()) -> any().
+start(Host, Opts) ->
+    mod_global_distrib_utils:start(?MODULE, Host, Opts, fun start/0).
+
+-spec stop(Host :: ejabberd:lserver()) -> any().
+stop(Host) ->
+    mod_global_distrib_utils:stop(?MODULE, Host, fun stop/0).
+
+-spec deps(Host :: ejabberd:server(), Opts :: proplists:proplist()) -> gen_mod:deps_list().
+deps(Host, Opts) ->
+    mod_global_distrib_utils:deps(?MODULE, Host, Opts, fun deps/1).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([RefreshInterval]) ->
+    ?DEBUG("refresher starting with interval ~p~n", [RefreshInterval]),
+    schedule_refresh(RefreshInterval),
+    {ok, RefreshInterval}.
+
+handle_call(Request, From, State) ->
+    ?ERROR_MSG("issue=unknown_call request=~p from=~p", [Request, From]),
+    {reply, {error, unknown_request}, State}.
+
+handle_cast(Request, State) ->
+    ?ERROR_MSG("issue=unknown_cast request=~p", [Request]),
+    {noreply, State}.
+
+handle_info(refresh, Interval) ->
```

Comment:
I think we'd do well to handle other, unexpected messages here. While call and casts really shouldn't happen without us coding the wrong logic, normal messages could be generated by random things in runtime over which we have no control.

Reply:
Of course, I agree, but what should I do then? Not log it? What implementation of handle_call, handle_cast should I provide? Or maybe this whole module shouldn't be a gen_server as it does not expect any calls or casts? Please elaborate :)

Reply:
The comment is not about `handle_cast` & `handle_call`, it's about providing a catch-all for `handle_info`.

Reply:
Ahh, okey